asoman/anpfGREdecap (#3972)
openconfig/featureprofiles: feature/policy_forwarding/otg_tests/mpls_gre_ipv4_decap_test/README.md

* IPV4 and IPV6 addresses

* MTU (default 9216)

* LACP Member link configuration

* Lag id

* LACP (default: period short)

* Carrier-delay (default up:3000 down:150)

* Statistics load interval (default:30 seconds)

### Routing

* MPLSoGRE decapsulation prefix range must be configurable on the device.  MPLSoGRE traffic within prefix ranges must be processed by the device.

* Static mapping of MPLS label to an egress nexthop must be configurable. Egress nexthop is based on the MPLS label/ static LSP.

* MPLS label for a single egress VLAN interface must be unique for decapsulated traffic:
    * IPV4 traffic
    * IPV6 traffic
    * Multicast traffic

* ECMP (Member links in Aggregate1) based on:
    * inner IP packet header  AND/OR
    * MPLS label, Outer IP packet header 

* Inner packet TTL and DSCP must be preserved during decap of traffic from MPLSoGRE to IP traffic

### MPLS Label 

* Entire Label block must be reallocated for static MPLS

* Labels from start/end/mid ranges must be usable and configured corresponding to MPLSoGRE decapsulation

### Multicast

* Multicast traffic must be decapsulated and sent out with L2 header based on the multicast payload address.

> ## NOTE: All test cases expected to meet following requirements and need not be a test pass criteria

Reviewer: ```suggestion
## NOTE: All test cases expected to meet following requirements even though they are not explicitly validated in the test.
```

* Use 64, 128, 256, 512, 1024.. MTU bytes frame size
* Increase the number of VLANs under PortChannel1 with traffic running until the maximum possible VLANs available on the device

Verify:
* Egress routes are programmed and LACP bundles are up without any errors, chassis alarms or exception logs
* There is no recirculation (iow, no impact to line rate traffic. No matter how the port allocation is done) of traffic
* All traffic received on Port-Channel2 and Port-Channel3 gets decapsulated and forwarded as IPV4/IPV6 unicast on the respective egress interfaces under Port-Channel1
* No packet loss when forwarding with counters incrementing corresponding to traffic
* Traffic equally load-balanced across member links of the port channel
* Header fields are as expected without any bit flips
* Inner payload DSCP and TTL values are not altered by the device
* Device can achieve the maximum interface scale on the device
* Entire static label range is usable and functional by sending traffic across the entire label range

## OpenConfig Path Coverage
TODO: Finalize and update the below paths after the review and testing on any vendor device.

### JSON Format

> ```

Reviewer: ```suggestion
```json
```

                  "name": "LA_POP 169.254.100.138 in:40422 out:pop"
                },
                                {
                  "config": {
                    "name": "LA_POP 2600:2d00:0:1:4000:15:7d:e1ba in:69128 out:pop"
                  },
                  "egress": {
                    "config": {
                      "incoming-label": 69128,
                      "next-hop": "2600:2d00:0:1:4000:15:7d:e1ba",
                      "payload-type": "IPV6"
                    }
                  },
                  "name": "LA_POP 2600:2d00:0:1:4000:15:7d:e1ba in:69128 out:pop"
                },
                {
                  "config": {
                    "name": "LA_POP 2600:2d00:0:1:4000:15:7d:e20a in:69775 out:pop"
                  },
                  "egress": {
                    "config": {
                      "incoming-label": 69775,
                      "next-hop": "2600:2d00:0:1:4000:15:7d:e20a",
                      "payload-type": "IPV6"
                    }
                  },
                  "name": "LA_POP 2600:2d00:0:1:4000:15:7d:e20a in:69775 out:pop"
                },
              ]
            }
          }
        },
```

> ## Telemetry Path Coverage
> ####
> 
> ```

Reviewer: You can leave out the basic interface counters as those should be covered in the [interface featureprofiles tests](https://github.com/openconfig/featureprofiles/tree/main/feature/interface).  Instead, just include the telemetry which is unique to your feature.  (the policy-forwarding counters)

But you do need to update the heading name and yaml syntax here to pass the github check for formatting.

```suggestion
## OpenConfig Path and RPC Coverage

```yaml
paths:
    # Telemetry for GRE decap rule
   /network-instances/network-instance/policy-forwarding/policies/policy/policy-counters/state/out-pkts:
   /network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/state/matched-pkts:
   /network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/state/sequence-id:

```

# PF-1.5 - MPLSoGRE IPV4 decapsulation of IPV4/IPV6 payload 

## Summary
This test verifies MPLSoGRE decapsulation of IP traffic using static MPLS LSP configuration. MPLSoGRE Traffic on ingress to the DUT is decapsulated and IPV4/IPV6 payload is forwarded towards the IPV4/IPV6 egress nexthop.

## Testbed type
* [`featureprofiles/topologies/atedut_8.testbed`](https://github.com/openconfig/featureprofiles/blob/main/topologies/atedut_8.testbed)

## Procedure
### Test environment setup

```

> DUT has an ingress and 2 egress EtherChannels.

Author: Done

                         |         | --eBGP-- | ATE Ports 3,4 |
    [ ATE Ports 1,2 ]----|   DUT   |          |               |
                         |         | --eBGP-- | ATE Port 5,6  |
```

Test uses aggregate 802.3ad bundled interfaces (Port Channel).

* Ingress Ports: Port-Channel2 and Port-Channel3
    * Port-Channel2 (ATE Ports 3,4) and Port-Channel3 (ATE Ports 5,6) are used as the source ports for encapsulated traffic.

* Egress Ports: Port-Channel1
    * Traffic is forwarded (egress) on Port-Channel1 (ATE Ports 1,2) .

#### Configuration

> #### Port-Channel1 is the egress port having following configuration:

Author: Done

> #### Ten or more subinterfaces (customer) with different VLAN-IDs

Author: Done

    * inner IP packet header  AND/OR
    * MPLS label, Outer IP packet header 

* Inner packet TTL and DSCP must be preserved during decap of traffic from MPLSoGRE to IP traffic

### MPLS Label 

* Entire Label block must be reallocated for static MPLS

* Labels from start/end/mid ranges must be usable and configured corresponding to MPLSoGRE decapsulation

### Multicast

* Multicast traffic must be decapsulated and sent out with L2 header based on the multicast payload address.

## PF-1.5.1: Verify MPLSoGRE decapsulate action for IPv4 and IPV6 payload
Generate traffic on ATE Ports 3,4,5,6 having:
* Outer source address: random combination of 1000+ IPV4 source addresses
* Outer destination address: Traffic must fall within the configured IPV4 unicast decap prefix range for MPLSoGRE traffic on the device
* MPLS Labels: Configure streams that map to every egress interface by having associated IPV4/IPV6/Multicast static MPLS labels in the MPLSoGRE header
* Inner payload: 
  * Both IPV4 and IPV6 unicast payloads, with random source address, destination address, TCP/UDP source port and destination ports
* Use 64, 128, 256, 512, 1024.. MTU bytes frame size

Verify:

> * Egress routes are programmed and LACP bundles are up without any errors, chassis alarms or exception logs

Reviewer: How should all these things verified? You have to specify exactly how to verify this in custom instructions here in this test, or test helpers that have already been defined to do these things.

> * There is no recirculation (iow, no impact to line rate traffic. No matter how the port allocation is done) of traffic

Reviewer: I think you have to remove this from the functional test, and create a subtest for performance. The performance test can send high rate traffic and look for packet loss.

Author: Added a note section as discussed. Need not be a test pass criteria

* All traffic received on Port-Channel2 and Port-Channel3 gets decapsulated and forwarded as IPV4/IPV6 unicast on the respective egress interfaces under Port-Channel1

> * No packet loss when forwarding with counters incrementing corresponding to traffic

Reviewer: which counters do you want to verify?  It's not strictly necessary imo to check basic interface counters.   Do you want to check policy-forwarding counters? (ie: how many packets are decapped?)

You can look these up at:

https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-network-instance.html#network-instances-network-instance-policy-forwarding-policies-policy-rules-rule-state-matched-pkts

Author: Added policy counters in the telemetry section

Reviewer: thanks I see them at the end of the file now

* Traffic equally load-balanced across member links of the port channel
  * Based on inner payload
  * Based on outer payload
  * Based on inner and outer payload
* Header fields are as expected without any bit flips
* Remove and add IPV4 configs and verify that there is no impact on IPV6 traffic

> * Remove and add IPV6 configs and verify that there is no impact on IPV4 traffic

Author: Done

* Inner payload: 
  * Multicast traffic with random source address, TCP/UDP header with random source and destination ports
* Use 64, 128, 256, 512, 1024.. MTU bytes frame size

Verify:

* Egress routes are programmed and LACP bundles are up without any errors, chassis alarms or exception logs
* There is no recirculation (iow, no impact to line rate traffic. No matter how the port allocation is done) of traffic
* All traffic received on Port-Channel2 and Port-Channel3 gets decapsulated and forwarded as multicast traffic on the respective egress interfaces under Port-Channel1
* No packet loss when forwarding with counters incrementing corresponding to traffic
* Traffic equally load-balanced across member links of the port channel
  * Based on inner payload
  * Based on outer payload
  * Based on inner and outer payload
* Header fields are as expected without any bit flips
* Verify that multicast L2 rewrite/egress headers are correct based on the multicast payload IPV4 destination address 

## PF-1.5.3: Verify MPLSoGRE DSCP/TTL preserve operation 
Generate traffic on ATE Ports 3,4,5,6 having:

> * Outer source address: random combination of 1000+ IPV4 source addresses

Author: Done, seems like idea is to prevent possible traffic leaks. i am more inclined not to have this as a hard requirement.

* Outer destination address: Traffic must fall within the configured IPV4 unicast decap prefix range for MPLSoGRE traffic on the device
* MPLS Labels: Configure streams that map to every egress interface by having associated IPV4/IPV6/Multicast static MPLS labels in the MPLSoGRE header
* Inner payload with all possible DSCP range 0-56 : 
  * Both IPV4 and IPV6 unicast payloads, with random source address, destination address, TCP/UDP source port and destination ports
  * Multicast traffic with random source address, TCP/UDP header with random source and destination ports
* Use 64, 128, 256, 512, 1024.. MTU bytes frame size

Verify:

* Egress routes are programmed and LACP bundles are up without any errors, chassis alarms or exception logs

> * There is no recirculation (iow, no impact to line rate traffic. No matter how the port allocation is done) of traffic

Reviewer: Split the  performance test into a separate subtest.  

In the functional subtest, you'll have to specify how to verify the packets are decapped on the ATE. (ATE IP packet counters good enough? Or do you want to decode packets and verify addresses? etc?). 

For the performance test, we don't want to decode millions of packets which would be expensive in terms of execution time.    So describe the performance flow and how the verification should  be done in detail.  ie: define the flow and verification such as:  Send 10M packets at 99Gbps rate and verify 10M packets are received.

Author: Added a note section as discussed. Need not be a test pass criteria

* No packet loss when forwarding with counters incrementing corresponding to traffic
* Traffic equally load-balanced across member links of the port channel
* Header fields are as expected without any bit flips
* Inner payload DSCP and TTL values are not altered by the device

## PF-1.5.4: Verify IPV4/IPV6 nexthop resolution of decap traffic
Generate traffic on ATE Ports 3,4,5,6 having:
* Outer source address: random combination of 1000+ IPV4 source addresses
* Outer destination address: Traffic must fall within the configured IPV4 unicast decap prefix range for MPLSoGRE traffic on the device
* MPLS Labels: Configure streams that map to every egress interface by having associated IPV4/IPV6/Multicast static MPLS labels in the MPLSoGRE header
* Inner payload with all possible DSCP range 0-56 : 
  * Both IPV4 and IPV6 unicast payloads, with random source address, destination address, TCP/UDP source port and destination ports
  * Multicast traffic with random source address, TCP/UDP header with random source and destination ports
* Use 64, 128, 256, 512, 1024.. MTU bytes frame size
* Clear ARP entries and IPV6 neighbors on the device

Verify:

> * Device must resolve the ARP and IPV6 neighbors upon receiving traffic

Reviewer: How is this verified?

Author: Added a note section as discussed. Need not be a test pass criteria

> * No packet loss when forwarding with counters incrementing corresponding to traffic

Reviewer: need to specify which counters to verify.  OTG?  or DUT or both?

Author: Added a note section as discussed. Need not be a test pass criteria

  * Multicast traffic with random source address, TCP/UDP header with random source and destination ports
* Use 64, 128, 256, 512, 1024.. MTU bytes frame size
* Increase the number of VLANs under PortChannel1 with traffic running until the maximum possible VLANs available on the device

Verify:
* Egress routes are programmed and LACP bundles are up without any errors, chassis alarms or exception logs
* There is no recirculation (iow, no impact to line rate traffic. No matter how the port allocation is done) of traffic
* All traffic received on Port-Channel2 and Port-Channel3 gets decapsulated and forwarded as IPV4/IPV6 unicast on the respective egress interfaces under Port-Channel1
* No packet loss when forwarding with counters incrementing corresponding to traffic
* Traffic equally load-balanced across member links of the port channel
* Header fields are as expected without any bit flips
* Inner payload DSCP and TTL values are not altered by the device
* Device can achieve the maximum interface scale on the device
* Entire static label range is usable and functional by sending traffic across the entire label range

> ## OpenConfig Path Coverage

Author: Done

              {
                "config": {
                  "policy-id": "pf-decap-range"
                },
                "entry-groups": {
                  "entry-group": [
                    {
                      "action": {
                        "config": {
                          "decapsulate-gre": true
                        }
                      },
                      "config": {
                        "group-id": 0
                      },
                      "group-id": 0,
                      "matches": {
                        "match": [
                          {
                            "config": {
                              "sequence-id": 0
                            },
                            "ipv4": {
                              "config": {
                                "destination-address": "10.1.157.160/28"
                              }
                            },
                            "sequence-id": 0
                          }
                        ]
                      }
                    }
                  ]
                },
                "policy-id": "pf-decap-range"
              }

>         }

Reviewer: refactor to use valid OC paths.

Example canonical OC

### policy-forwarding policy
```json
"network-instances": {
  "network-instance": {
    "DEFAULT": {
       "name": "default",
       "policy-forwarding": {
         "policies": {
           "policy": [
              {
                "config": {
                  "policy-id": "decap MPLS in GRE"
                },
                "rules": {
                  "rule": [
                    {
                      "config": {
                        "sequence-id": 1
                      },
                      "ipv4": {
                        "config": {
                          "destination-address": "169.254.125.155/32",
                          "protocol": "IP"
                        },
                        }
                    },
                    "action": {
                        "decapsulate-gre": true
                        }
                      },
                      "sequence-id": 1
                    }
                  ]
                },
              }
           ]
         }
       }
    }
  }
}
```

Author: Done.

          "mpls": {
          "global": {
            "interface-attributes": {
              "interface": [
                {
                  "config": {
                    "interface-id": "Port-Channel4",
                    "mpls-enabled": false
                  },
                  "interface-id": "Port-Channel4"
                }
              ]
            }
          },
          "lsps": {
            "static-lsps": {
              "static-lsp": [
                {
                  "config": {
                    "name": "LA_POP 169.254.1.138 in:40571 out:pop"
                  },
                  "egress": {
                    "config": {
                      "incoming-label": 40571,
                      "next-hop": "169.254.1.138",
                      "payload-type": "IPV4"

>                     }

Reviewer: refactor to use real OC paths.  I searched the repo for `static lsp` and found  a static-lsp test here:

https://github.com/openconfig/featureprofiles/blob/d6cfda756448f43b2b51cad9b768c4003b1b763f/feature/mpls/otg_tests/static_lsp/README.md?plain=1#L1

Author: Done

                    }
                  },
                  "name": "LA_POP 2600:2d00:0:1:4000:15:7d:e1ba in:69128 out:pop"
                },
                {
                  "config": {
                    "name": "LA_POP 2600:2d00:0:1:4000:15:7d:e20a in:69775 out:pop"
                  },
                  "egress": {
                    "config": {
                      "incoming-label": 69775,
                      "next-hop": "2600:2d00:0:1:4000:15:7d:e20a",
                      "payload-type": "IPV6"
                    }
                  },
                  "name": "LA_POP 2600:2d00:0:1:4000:15:7d:e20a in:69775 out:pop"
                },
              ]
            }
          }
        },
```

## OpenConfig Path and RPC Coverage

```yaml
paths:
    # Telemetry for GRE decap rule    
    /network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/state/matched-pkts:
    /network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/state/matched-octets:
    
    # Config paths for GRE decap
    /network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/action/config/decapsulate-gre:

> interfaces/interface/state/counters/in-discards

Author: Done

interfaces/interface/subinterfaces/subinterface/ipv4/state/counters/out-error-pkts
interfaces/interface/subinterfaces/subinterface/ipv4/state/counters/out-forwarded-pkts
interfaces/interface/subinterfaces/subinterface/ipv4/state/counters/out-multicast-pkts
interfaces/interface/subinterfaces/subinterface/ipv4/state/counters/out-pkts

interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/in-discarded-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/in-error-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/in-forwarded-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/in-multicast-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/in-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/out-discarded-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/out-error-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/out-forwarded-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/out-multicast-pkts
interfaces/interface/subinterfaces/subinterface/ipv6/state/counters/out-pkts

> network-instances/network-instance/policy-forwarding/policies/policy/policy-counters/state/out-pkts

Author: Done

> network-instances/network-instance/policy-forwarding/policies/policy/policy-counters/state/out-pkts

Author: Done
